handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -307,20 +307,39 @@ public void operationComplete(Future<SslContext> future) throws Exception {
 
     private void replaceHandler(ChannelHandlerContext ctx, Selection selection) {
         SSLEngine sslEngine = null;
+        SslHandler handler = null;
         this.selection = selection;
         try {
             sslEngine = selection.context.newEngine(ctx.alloc());
-            ctx.pipeline().replace(this, SslHandler.class.getName(), selection.context.newHandler(sslEngine));
+            handler = SslContext.newHandler(sslEngine);
+            replaceHandler(ctx, selection.hostname, selection.context, SslHandler.class.getName(), handler);
         } catch (Throwable cause) {
             this.selection = EMPTY_SELECTION;
             // Since the SslHandler was not inserted into the pipeline the ownership of the SSLEngine was not
             // transferred to the SslHandler.
             // See https://github.com/netty/netty/issues/5678
             ReferenceCountUtil.safeRelease(sslEngine);
+
+            if (handler != null) {
+                handler.sslCloseFuture().cancel(true);
```

Comment:
The `sslCloseFuture` may be a bit under defined but I'm not sure it is meaningful if the handshake never completes. Also it looks we only complete it if the SSL connection do a "graceful" shutdown (e.g. we get a shutdown from the peer).